### PR TITLE
[Feature] add eva02 backbone

### PR DESCRIPTION
**Pull request by zzc98**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add the EVA02 backbone.
paper: https://arxiv.org/abs/2303.11331
code: https://github.com/baaivision/EVA/tree/master/EVA-02

## Modification

- Add new backbone network EVA02
- Complete the network implementation and configuration files
- Complete the weight conversion according to the official repository
- Align the accuracy on the ImageNet1k validation set

## Use cases

```python
model = dict(
    type='ImageClassifier',
    backbone=dict(
        type='EVA02',
        arch='base',
        img_size=448,
        sub_ln=True,
        out_type='avg_featmap'),
    neck=None,
    head=dict(
        type='LinearClsHead',
        num_classes=1000,
        in_channels=768,
        loss=dict(
            type='LabelSmoothLoss', label_smooth_val=0.1, mode='original'),
    ),
    init_cfg=[
        dict(type='TruncNormal', layer='Linear', std=.02),
        dict(type='Constant', layer='LayerNorm', val=1., bias=0.),
    ],
    train_cfg=dict(augments=[
        dict(type='Mixup', alpha=0.8),
        dict(type='CutMix', alpha=1.0)
    ])
)
```

More details can be found in `configs/eva02/README.md`.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpretrain/models/backbones/eva02.py [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+        x = self.norm2(x)
+        x = self.mlp(x)
+        x = self.drop_path(x)
+        x = inputs + x
+
+        return x
+
+
+@MODELS.register_module()
+class EVA02(BaseModule):
```
Comment (Ezra-Yu) [inline mmpretrain/models/backbones/eva02.py:345]: why not Inherit ViT? note this [PR](https://github.com/open-mmlab/mmclassification/pull/1447/files)

File: mmpretrain/models/backbones/eva02.py [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+        self.proj = nn.Linear(embed_dims, embed_dims, bias=proj_bias)
+        self.proj_drop = nn.Dropout(proj_drop)
+
+        self.with_cls_token = with_cls_token
+        self.rope = rope
+
+    def forward(self, x, patch_resolution):
+        B, N, _ = x.shape
+
+        if self.q_bias is not None:
```
Comment (Ezra-Yu) [inline mmpretrain/models/backbones/eva02.py:215]: this section of code would be exe every iter. Can it be put in __init__.

File: mmpretrain/models/backbones/eva02.py [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+        inputs = x
+        x = x.reshape(batch, num_heads, num_patches, -1, 2)
+        x1, x2 = x.unbind(dim=-1)
+        x = torch.stack((-x2, x1), dim=-1)
+        x = x.reshape(batch, num_heads, num_patches, dim)
+
+        return inputs * self.freqs_cos + x * self.freqs_sin
+
+
+class SwiGLU(BaseModule):
```
Comment (Ezra-Yu) [inline mmpretrain/models/backbones/eva02.py:94]: `SwiGLU` -> `SwiGLUFFN`

File: mmpretrain/models/backbones/eva02.py [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+
+        x = (attn @ v).transpose(1, 2).reshape(B, N, -1)
+
+        x = self.proj(x)
+        x = self.proj_drop(x)
+
+        return x
+
+
+class Block(BaseModule):
```
Comment (Ezra-Yu) [inline mmpretrain/models/backbones/eva02.py:254]: EVA02EndcoderLayer？

File: mmpretrain/models/backbones/eva02.py [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+from mmcv.cnn.bricks.drop import build_dropout
+from mmcv.cnn.bricks.transformer import PatchEmbed
+from mmengine.model import BaseModule, ModuleList
+from mmengine.model.weight_init import trunc_normal_
+
+from mmpretrain.registry import MODELS
+from ..utils import build_norm_layer, resize_pos_embed, to_2tuple
+
+
+class RotaryEmbedding(BaseModule):
```
Comment (Ezra-Yu) [inline mmpretrain/models/backbones/eva02.py:18]: this module could be mv into `mmpretain/models/utils/postion_encoding.py`. unit test is needed.

**Review by Ezra-Yu — COMMENTED** [review]
please rm `_base_.models.eva` folder and put the models cfg into every file of config/eva02/eva02_xxxx.py.

**Review by Ezra-Yu — APPROVED** [review]
LGTM.

File: configs/eva02/metafile.yml [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        - Rotary Position Embedding
+        - Sub Layer Normalization
+        - SwiGLU
+    Paper:
+      Title: 'EVA-02: A Visual Representation for Neon Genesis'
+      URL: https://arxiv.org/abs/2303.11331
+    README: configs/eva02/README.md
+
+Models:
+  - Name: eva02-tiny-p14_pre_in21k
```
Comment (fangyixiao18) [inline configs/eva02/metafile.yml:14]: rename like `vit-tiny-p14_eva02-pre_in21k`, check all `Name` in metafile

File: mmpretrain/models/backbones/eva02.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        x = self.norm2(x)
+        x = self.mlp(x)
+        x = self.drop_path(x)
+        x = inputs + x
+
+        return x
+
+
+@MODELS.register_module()
+class EVA02(VisionTransformer):
```
Comment (fangyixiao18) [inline mmpretrain/models/backbones/eva02.py:200]:
```suggestion
class ViTEVA02(VisionTransformer):
```

File: mmpretrain/models/backbones/eva02.py
Comment (fangyixiao18) [inline mmpretrain/models/backbones/eva02.py:1]: filename: `vit_eva02.py`

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: configs/eva02/eva02_base_p14_8xb16_in1k.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
+_base_ = [
+    '../_base_/datasets/imagenet_bs16_eva_448.py',
+    '../_base_/schedules/imagenet_bs2048_AdamW.py',
+    '../_base_/default_runtime.py'
+]
+
+model = dict(
+    type='ImageClassifier',
+    backbone=dict(
+        type='EVA02',
```
Comment (fangyixiao18) [inline configs/eva02/eva02_base_p14_8xb16_in1k.py:10]:
```suggestion
        type='ViTEVA02',
```

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]
config names

File: docs/en/api/models.rst [hunk 9]
```diff
@@ -189,6 +189,7 @@ Backbones
    VisionTransformer
    ViTSAM
    XCiT
+   EVA02
```
Comment (fangyixiao18) [inline docs/en/api/models.rst:192]:
```suggestion
   ViTEVA02
```

**Review by fangyixiao18 — APPROVED** [review]
